crates/red_knot_python_semantic/src/module_name.rs

```diff
@@ -186,6 +187,45 @@ impl ModuleName {
         self.0.push('.');
         self.0.push_str(other);
     }
+
+    /// Appends a name to `self`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ruff_python_ast::name::Name;
+    /// use red_knot_python_semantic::ModuleName;
+    ///
+    /// let mut module_name = ModuleName::new_static("foo").unwrap();
+    /// module_name.append(&Name::new("bar"));
+    /// assert_eq!(&module_name, "foo.bar");
+    /// ```
+    pub fn append(&mut self, name: &Name) {
+        self.0.push('.');
+        self.0.push_str(name);
+    }
+
+    /// Returns an iterator of this module name and all of its parent modules.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use red_knot_python_semantic::ModuleName;
+    ///
+    /// assert_eq!(
+    ///     ModuleName::new_static("foo.bar.baz").unwrap().parents().collect::<Vec<_>>(),
+    ///     vec!["foo.bar.baz", "foo.bar", "foo"],
+    /// );
+    /// ```
+    #[must_use]
+    pub fn parents(&self) -> impl Iterator<Item = &str> {
+        let mut name = Some(self.as_str());
+        std::iter::from_fn(move || {
+            let result = name;
+            name = name.and_then(|n| n.rsplit_once('.').map(|(p, _)| p));
+            result
+        })
```

Comment:
Nit: I'd call this `ancestors` similar to `PathBuf::ancestors`. I'm further leaning towards naming it `ancestor_components` because the method returns a `&str`, unlike `ModuleName::Parent` that returns a `&str` or change the return type to `ModuleName` (or we introduce a `ModuleNameRef` if we're concerned about the potential allocation).

Nit2: I think we can simplify this by calling `ModuleName::parent` (Only true if you change the return type)

```
std::iter::successors(Some(self.clone()), Self::parent)
```

Reply:
I like that `successors` formulation.  At least at the moment, the extra allocations aren't an issue since we were already doing a `clone` over in the semantic index builder.  So this change just moves where the allocation occurs.